### PR TITLE
fix(notifications): stock-format .notify files (#134)

### LINKED ISSUE #134: [BUG] Notifications written to the `unraid` channel crash Unraid's PHP notify parser (phplog fatal flood); ENOSPC leaves 0-byte "invalid" notifications
<h2>Bug Description</h2>
<p>Hi there, I hit an odd corner case today but thought I'd share the issue and what I think the fix could be. The agent writes notification files directly into <code>/tmp/notifications/unread/</code>
 in a format that Unraid's stock PHP notification parser cannot parse  
every webGui notification poll then throws a PHP fatal, flooding <code>/var/log/phplog</code> (a log with no logrotate coverage, on the 128MB <code>/var/log</code> tmpfs). Additionally, if the filesystem is full when a dispatch occurs, the agent leaves 0-byte <code>.notify</code>
 files behind, which the Unraid UI renders as perpetual "This 
notification is invalid and cannot be displayed!" warnings re-stamped 
with the current time on every page load.</p>
<p>Both defects share one root cause: the <code>unraid</code> alert channel writes files itself instead of using (or exactly replicating) <code>/usr/local/emhttp/webGui/scripts/notify</code>.</p>
<h2>Environment</h2>
<p><strong>Unraid Version:</strong> 7.3.1
<strong>Plugin Version:</strong> 2026.06.07
<strong>Architecture:</strong> x86_64</p>
<h2>Hardware Configuration</h2>
<p>Not hardware-related (notification/alerting pipeline), but for completeness:</p>
<ul>
<li><strong>CPU:</strong> AMD Ryzen 5 3600</li>
<li><strong>Disk Controller:</strong> 2× LSI SAS2008 (Fusion-MPT SAS-2) HBAs + AMD 400-series chipset SATA</li>
<li><strong>GPU:</strong> NVIDIA GeForce GT 710</li>
<li><strong>UPS:</strong> PPC offline UPS via NUT</li>
<li><strong>Network:</strong> Intel I211 + dual-port Intel 82571EB, bonded (bond0)</li>
<li><strong>Other Relevant Hardware:</strong> N/A</li>
</ul>
<h2>Steps to Reproduce</h2>
<ol>
<li>Let any alert rule dispatch to the <code>unraid</code> channel — the agent writes e.g. <code>/tmp/notifications/unread/20260703-152000-Management_Agent_Alert.notify</code> directly.</li>
<li>Run <code>/usr/local/emhttp/plugins/dynamix/scripts/notify get</code> (the same code path the webGui notification poller uses).</li>
<li>Observe a PHP fatal at line 296 (exit code 255). Each UI poll repeats it, appending to <code>/var/log/phplog</code>.</li>
<li>Remove the agent-written file and run the same command: it returns <code>[]</code> with exit 0.</li>
<li>(0-byte variant) Cause ENOSPC on <code>/tmp</code> (RAM-backed rootfs full) while a dispatch fires: a 0-byte <code>.notify</code> file is left in <code>unread/</code>  the agent logs the write failure but never removes the empty file it created.</li>
</ol>
<h2>Expected Behavior</h2>
<ul>
<li>Notification files written by the agent are parseable by both stock consumers (the legacy PHP <code>notify</code> script and unraid-api), like files created via <code>notify</code> script itself.</li>
<li>A failed dispatch (e.g. ENOSPC) leaves no partial/empty file behind.</li>
</ul>
<h2>Actual Behavior</h2>
<ul>
<li>Every webGui notification poll fatals on agent-written files: <code>/var/log/phplog</code> accumulated ~189,000 fatals / 75MB (of the 128MB <code>/var/log</code>
 tmpfs) on this system. The loop first armed with the agent's very first
 notification (a plugin-update notice) and re-arms with any new one.</li>
<li>A rootfs-full incident left 15 zero-byte <code>.notify</code> husks (each matching a <code>Failed to create notification ... no space left on device</code> log line 1:1). The Unraid UI masks each invalid file as a WARNING with a <code>new Date()</code> timestamp, so they look like brand-new warnings on every page load, indefinitely.</li>
</ul>
<p>The format differences vs. the stock <code>notify</code> script:</p>

  | stock notify script | agent-written file
-- | -- | --
timestamp= | unquoted unix epoch (PHP reader does date('Y-m-d H:i', $timestamp)) | quoted datetime string "2026-07-03 15:20:00" → fatal
field order | timestamp first | event first
filename | hyphens/spaces stripped via safe_filename() | hyphens kept
archive copy | always written first | never written

<p>Example agent-written file (verbatim):</p>
<pre><code class="language-ini">event="Management Agent Alert"
subject="Resolved: Agent data source degraded"
description="Alert rule 'Agent data source degraded' resolved"
importance="warning"
timestamp="2026-07-03 15:20:00"
link=""
</code></pre>
<h2>API Endpoint (if applicable)</h2>
<p>Not an API bug but <code>GET /api/v1/alerts/history</code> 
corroborated the accounting perfectly: 28 dispatch events (2 firing, 26 
resolved) matching the 13 written + 15 failed files exactly. (Side 
observation: resolve events dispatch at <code>warning</code> importance 
with no cooldown 26 resolve notifications in ~100 minutes during a 
flap. Normal severity for resolves and/or hysteresis would be a nice 
improvement.)</p>
<h2>Logs</h2>
<p>From <code>/var/log/unraid-management-agent.log</code>:</p>
<pre><code>2026/07/03 01:16:15 ERROR: Failed to create notification: write /tmp/notifications/unread/20260703-011615-Management_Agent_Alert.notify: no space left on device
2026/07/03 01:16:15 ERROR: Alerting: Failed to dispatch to channel unraid for rule subsystem-degraded: failed to create notification: write /tmp/notifications/unread/20260703-011615-Management_Agent_Alert.notify: no space left on device
2026/07/03 15:20:00 Created notification: 20260703-152000-Management_Agent_Alert.notify
</code></pre>
<p>From <code>/var/log/phplog</code> (repeated ~189,000 times):</p>
<pre><code>PHP Fatal error:  Uncaught TypeError: date(): Argument #2 ($timestamp) must be of type ?int, string given in /usr/local/emhttp/plugins/dynamix/scripts/notify:296
  thrown in /usr/local/emhttp/plugins/dynamix/scripts/notify on line 296
</code></pre>
<h2>Screenshots</h2>
<p>N/A UI symptom is the Notifications tray showing, for each 0-byte 
file: "This notification is invalid and cannot be displayed! For 
details, see the logs and the notification file at 
/tmp/notifications/unread/&lt;name&gt;.notify", timestamped with the 
moment the page was loaded.</p>
<h2>Additional Context</h2>
<ul>
<li>First observed armed on 2026-06-22 (agent's first notification; 
first phplog fatal one second later); rediscovered during a 2026-07-03 
rootfs-full incident that produced the 15 husks. Full forensic evidence 
(file inventories, 1:1 log correlation, UI code references) available on
 request.</li>
<li>Workaround: delete agent-format files from <code>/tmp/notifications/unread/</code>, truncate <code>/var/log/phplog</code>, and add a logrotate rule for phplog. Any new agent notification re-arms the fatal loop.</li>
<li>The <code>new Date()</code> masking of invalid files is arguably an unraid-api issue too, but the agent can avoid creating such files entirely.</li>
</ul>
<h2>Possible Solution</h2>
<ol>
<li>For the <code>unraid</code> channel, shell out to the stock script <code>notify -e "&lt;event&gt;" -s "&lt;subject&gt;" -d "&lt;description&gt;" -i &lt;importance&gt;</code> or replicate its output exactly (unquoted epoch <code>timestamp=</code> first, <code>safe_filename()</code> naming, archive copy).</li>
<li>Write dispatches atomically: write to a temp file and <code>rename()</code> into <code>unread/</code>; on write error, <code>unlink</code> so ENOSPC can never leave a visible partial file.</li>
<li>Optional: dispatch resolve events at <code>normal</code> importance and add a cooldown/hysteresis to state-transition notifications.</li>
</ol>

### PR DESCRIPTION
**Pull request by DeeGeeGit**
Hiya! 👋 Thanks for taking a look at this. I hit the edge case I'd reported in #134 and since
I'd done some debugging, the code wasn't much of an extra step.

## Description

Notification files written by the agent crash Unraid's stock PHP notification parser and, under
ENOSPC, leave 0-byte `.notify` husks behind (#134). This PR makes `CreateNotification` write a
stock-compatible file format matching the webGui `notify` script's output, and makes every write
atomic.

The stock `notify get` parser (used by the webGui notification poll) blindly runs
`date($fmt, $val)` on the **first** `key=value` line of each file, with no key check. Agent files
started with `event="..."` (and carried a quoted datetime `timestamp="..."` further down), so
every poll threw a PHP `TypeError` fatal, flooding `/var/log/phplog` (~189k fatals / 75MB on my
system). Separately, `os.WriteFile` under ENOSPC creates the file before failing the
write, leaving 0-byte husks the UI flags as "invalid notification" with a fresh timestamp on
every page load.

All notification producers go through `CreateNotification`, so the alerting `unraid` channel,
`POST /api/v1/notifications`, watchdog, and plugin-update notices are all fixed at once.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Hardware compatibility fix
- [ ] Performance improvement
- [ ] Code refactoring (no functional changes)
- [x] Test additions/improvements

Marked non-breaking as the behavioral notes under Breaking Changes affect only edge-case titles
from arbitrary API callers (plus id/timestamp shape); every built-in producer and stock consumer
is unaffected. Happy to reclassify if you read the template's definition more strictly.

## Related Issues

Fixes #134

## Hardware Configuration (if applicable)

Not hardware-related (notification/alerting pipeline).

- **Unraid Version:** reproduced on 7.3.1, plugin 2026.06.07, x86_64

## Changes Made

All format details were verified line-by-line against the stock script's `build_ini_string()`,
`ini_decode_value()`, and `safe_filename()` in unraid/webgui.

- **Stock-compatible file body** (`daemon/services/controllers/notification.go`): unquoted
  unix-epoch `timestamp=` as the first line, stock field order (`timestamp`, `event`, `subject`,
  `description`, `importance`, then `link` in the unread copy only), backslash/double-quote
  escaping identical to `ini_encode_value()`, trailing newline.
- **Stock-style file names**: Go replica of `safe_filename()` applied to the title +
  `_<epoch>.notify` (e.g. `Management_Agent_Alert_1783092000.notify`), replacing the previous
  `20060102-150405-Title.notify` scheme. The pre-existing 50-char title cap and path-traversal
  guard are retained. The replica deliberately omits Go's `(?i)` flag: RE2 case-folds character
  classes before negation, which would keep U+212A/U+017F that stock's byte-oriented pattern
  strips (and A-Z is already inside the pattern's 0x20-0x5F range).
- **Archive copy written first**, best-effort, like stock (stock ignores archive write errors;
  the unread copy alone decides the call's outcome). Dismissed notifications now keep their
  archive history, matching stock behaviour. Archiving via the agent's API accordingly now
  removes the unread file and keeps that creation-time copy, like the stock `archive` verb (a
  rename would clobber it); files without a creation copy still fall back to the old rename, and
  if the existence check itself fails the error is surfaced and neither file is touched.
- **Atomic writes**: temp file (`.notify-tmp-*`, never matches `*.notify` globs) + `rename()`,
  with the temp removed on any failure — a failed dispatch can no longer leave a partial or
  empty file. This is deliberately stronger than stock, which also husks under ENOSPC.
- **Collector parser** (`daemon/services/collectors/notification.go`) now reads the stock format
  it writes (unquoted epoch, escaped values via an `ini_decode_value()` equivalent) while still
  parsing the legacy quoted-datetime format of files created by older agent versions. This also
  fixes timestamps for notifications created by the stock script itself, which previously fell
  back to file mtime. Parsed epochs are bounded to years [1970, 9999] and timestamps are stored
  as UTC, so every parse path renders the same Z-suffixed RFC3339 form regardless of host
  timezone and the cap alone bounds the year `MarshalJSON` validates. Out-of-range values fall
  back to file mtime instead of producing a `time.Time` that fails JSON marshaling and would
  otherwise break encoding of the whole notification list (REST/WS/MQTT).
- **CHANGELOG.md**: entry added under `## [Unreleased]` describing the fix.
- **Tests**: new golden-format, atomicity, archive-semantics, and parser tests; existing tests
  updated where they pinned the old format (details below).

## Testing Performed

- [x] Unit tests pass (`make test` — full suite; see sandbox note under Test Results)
- [x] Added new tests for new functionality
- [ ] Tested on actual Unraid system (my affected system from #134 is available for
      before/after validation prior to merge)
- [x] Verified affected API endpoints work correctly (unit level; no API signatures changed)
- [ ] Tested WebSocket events (n/a — no event shape changes)
- [ ] Tested with debug logging enabled
- [x] Regression testing (full suite, plus `go test -race` on the touched packages)

### Test Results

New/updated tests (table-driven where the repo convention applies):

- `TestCreateNotification_StockNotifyFormat` — byte-exact golden check of the unread file
  (epoch-first unquoted timestamp, stock field order, `\"`/`\\` escaping, trailing newline),
  `safe_filename()`-style naming, and the 0644 mode the web UI needs.
- `TestCreateNotification_WritesArchiveCopy` — archive copy present, identical minus the `link`
  line.
- `TestCreateNotification_UnreadFailureKeepsArchiveCopy` — pins the stock write order: archive
  first, retained when only the unread write fails.
- `TestCreateNotification_LongTitleCapped` — 50-char title cap retained.
- `TestCreateNotification_NoFileLeftOnWriteFailure` — failed dispatch leaves no file, no husk,
  no temp (skips as root, where directory permissions aren't enforced).
- `TestCreateNotification_ArchiveFailureDoesNotAbort` — archive write failure is tolerated like
  stock; unread copy still delivered.
- `TestArchiveNotification_KeepsCreationArchiveCopy` and
  `TestArchiveAllNotifications_KeepsCreationArchiveCopies` — archiving removes the unread file
  and preserves the creation-time archive copy instead of clobbering it.
- `TestArchiveOrRemove_AmbiguousCheckTouchesNothing` — a failed existence check on the archive
  destination returns the error without attempting a rename or remove; both files untouched.
- `TestWriteFileAtomic_CleansUpTempOnFailure` — atomic writer removes its temp on failure.
- `TestParseNotificationFile_StockFormat` — collector parses unquoted epoch, legacy quoted
  datetime, escaped values, falls back to the file's actual mtime for out-of-range epochs, pins
  the Z-suffixed UTC rendering with a host-independent literal, and stays JSON-marshalable at
  the epoch cap in any timezone (verified manually under `TZ=Asia/Tokyo` and
  `TZ=Pacific/Kiritimati`).
- `TestSafeFilename` — updated to the stock `safe_filename()` semantics, including the
  case-folding pitfall rows.
- `TestCreateNotification_SpecialCharactersInTitle` — tightened: titles whose sanitized form
  contains `..` (e.g. `Test...Dots`) are now explicitly expected to be rejected (see Breaking
  Changes).
- Test hygiene: the pre-existing importance-validation subtest and two api-package notification
  handler tests now redirect to temp dirs / skip as root — with the new archive-first write they
  would otherwise create real files under `/boot` when the suite runs as root on a dev host.

Every guarded behaviour was mutation- or red-tested: re-introducing the original bug
(event-first/quoted timestamp), removing the temp cleanup, disabling epoch parsing, and aborting
on archive failure each made exactly the guarding test fail; the archive-clobber,
timezone-boundary, and ambiguous-check tests were written failing against the defect first.

`gofmt` clean, `go vet` clean, `golangci-lint run --new-from-rev` reports 0 new issues,
`go test -race` clean on the touched packages (controllers, collectors, api).

Full-suite note: development ran in a network-isolated sandbox where two pre-existing MQTT tests
(`TestTestConnection`, `TestPackageLevelTestConnection_InvalidBroker`) fail because instant DNS
failure yields 0ms latency; they fail identically on a clean checkout in that sandbox and pass
in normal environments.

## **API Endpoints Tested:**

Unit level (handler and controller tests; no signatures changed):

- `POST /api/v1/notifications` — creation via handler tests, including default-importance path.
- `GET /api/v1/notifications` — parse/marshal path via collector tests (stock, legacy, and
  boundary-epoch inputs).
- `POST /api/v1/notifications/{id}/archive`, `.../unarchive`, `DELETE /api/v1/notifications/{id}`
  — controller tests, including the new keep-creation-archive-copy semantics.

- **Test Output/Results:**

```
ok  github.com/ruaan-deysel/unraid-management-agent/daemon/services/controllers
ok  github.com/ruaan-deysel/unraid-management-agent/daemon/services/collectors
ok  github.com/ruaan-deysel/unraid-management-agent/daemon/services/api
(full `go test ./...` green apart from the two documented sandbox-environment MQTT
failures above, which also fail on a clean checkout in that sandbox; -race green on
the three packages listed)
```

**Log Output (if relevant):**

```
(n/a — see issue #134 for the phplog fatals and agent log excerpts this fixes)
```

## Documentation

- [x] Code comments added/updated
- [ ] README.md updated (if needed) — not needed
- [ ] CLAUDE.md updated (if architecture changed) — no architecture change
- [ ] CONTRIBUTING.md updated (if contribution process changed) — n/a
- [ ] API documentation updated (if API changed) — no API shape change
- [ ] No documentation needed
- [x] CHANGELOG.md updated (entry under `## [Unreleased]`)

## **Breaking Changes:**

None for stock consumers. Four behavioral notes for consumers of the agent's own API:

1. Notification `id` values (the filename, as returned by `/api/v1/notifications` and used by
   archive/delete operations) change shape for newly created notifications:
   `20260703-152000-Management_Agent_Alert.notify` → `Management_Agent_Alert_1783092000.notify`.
   Existing files keep working; IDs remain opaque `.notify` filenames.
2. API `timestamp` fields change in two ways: stock-created notifications now use the epoch
   stored in the file instead of the near-identical mtime fallback, and newly created agent
   notifications now carry the correct instant (the old code wrote zoneless local wall time that
   the collector parsed as UTC, so on non-UTC hosts their API timestamps previously ran ahead or
   behind by the host's UTC offset). All timestamps are now rendered uniformly in UTC
   (Z-suffixed RFC3339) regardless of host timezone. Legacy-format agent files' timestamps parse exactly as
   before; quoted legacy values containing literal backslashes are now decoded with stock
   `stripslashes` semantics (the old reader left them intact), a niche case since the old
   writer never escaped values.
3. Titles whose sanitized form contains `..` (e.g. `Test...Dots`) are now rejected with an
   error: dots survive the stock `safe_filename()` replica (the old sanitizer stripped them),
   and the retained path-traversal guard rejects `..` anywhere in a filename — stricter than
   stock, which would create such files. All built-in producers use dot-free titles; only
   arbitrary API callers can hit this, and the rejection is now pinned by a test.
4. Two more edge-title outcomes change for arbitrary API callers, both pinned by tests: titles
   containing none of `[a-zA-Z0-9_-]` but only characters stock keeps (e.g. `@%^+.`) were
   previously rejected as empty-after-sanitize and now create notifications like stock would
   (a title starting with `.` yields a hidden leading-dot filename, which the agent's API lists
   but stock's `*.notify` glob skips); conversely, whitespace-only titles were previously
   accepted as underscore filenames and are now rejected as empty, because trimming happens
   before space conversion, matching stock's trim placement.

## **Migration Guide:**

None required. API consumers that treat notification IDs as opaque `.notify` filenames are
unaffected.

## Screenshots/Logs

N/A — UI symptoms, phplog excerpts, and agent log excerpts are documented in issue #134.

## Checklist

- [x] My code follows the project's coding standards
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings (`golangci-lint --new-from-rev`: 0 issues)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published (n/a — none)
- [x] I have read and followed the [CONTRIBUTING.md](../CONTRIBUTING.md) guidelines
- [x] No sensitive information (API keys, passwords, personal data) is included

## Hardware Compatibility Notes

N/A — not a hardware compatibility fix.

## Additional Notes

Intentional non-goals, kept out to stay minimal (happy to follow up on any):

- Stock's per-importance entity routing (browser/email/agents gating from `dynamix.cfg`) and its
  same-name dedup skip are not replicated; the agent still always writes the unread copy. As a
  consequence, two notifications for the same event within the same second overwrite each other
  (last write wins, atomically) where stock would skip the second entirely.
- `clean_subject()` HTML-entity stripping is not replicated (agent subjects never contain
  entities).
- Extra fsnotify `collect()` cycles from temp-file create events are accepted; temp names never
  match `*.notify`, so nothing ever parses them.

Related observations, not addressed here:

- Pre-existing: the alert dispatcher maps info-severity to importance `"normal"`
  (`daemon/services/alerting/dispatcher.go`), which `CreateNotification` rejects (it accepts
  `alert|warning|info`), so info-severity alerts to the `unraid` channel currently fail. Note
  also stock's vocabulary is `normal|warning|alert` — `"info"` is agent-only. Left untouched to
  keep this PR scoped to #134; can file separately.
- Pre-existing: `TestSanitizeFilenameCases` in `notification_additional_test.go` tests a local
  test-only helper (`sanitizeTestFilename`) that documents the old sanitization semantics; left
  untouched per "don't delete unrelated code", flagging it for a future cleanup.
- The issue's side observation about resolve-event severity/cooldown is not addressed here
  (explicitly optional in the issue).

---

Thanks a lot for taking the time to review this (and for the plugin itself!). Happy to adjust
anything, split it up further or anything else I can do to help. Cheers!

## Summary by CodeRabbit

* **New Features**
  * Notification `.notify` output is now stock-compatible INI-style `key=value`, with safer filename derivation (`event_<unix_epoch>.notify`) and improved archiving behaviour.
* **Bug Fixes**
  * Fixed crashes when parsing agent-created notifications by supporting both legacy and stock notification formats.
  * Improved `timestamp` handling (epoch bounds, consistent UTC/RFC3339 formatting, and safe fallbacks).
  * Atomic, two-phase writes prevent empty/partial files; archiving no longer overwrites existing archive copies.
* **Tests**
  * Added stock-format parsing/formatting tests, expanded atomic/archive/read-order assertions, and refined filename safety coverage (including root-run skips).

### CODE REVIEW

File: daemon/services/collectors/notification.go [hunk 1]
```diff
@@ -240,7 +241,19 @@ func (c *NotificationCollector) parseNotificationFile(path string, notifType str
...
+			// checked too: near the cap, east-of-UTC hosts render year 10000.
+			// (The epoch cap runs first so time.Unix cannot overflow.)
+			if epoch, err := strconv.ParseInt(value, 10, 64); err == nil && epoch >= 0 && epoch <= 253402300799 &&
+				time.Unix(epoch, 0).Year() <= 9999 {
+				notification.Timestamp = time.Unix(epoch, 0)
+				notification.FormattedTimestamp = notification.Timestamp.Format(time.RFC3339)
+			} else if ts, err := time.Parse("2006-01-02 15:04:05", value); err == nil {
 				notification.Timestamp = ts
 				notification.FormattedTimestamp = ts.Format(time.RFC3339)
 			}
```
[the remark this answers is not in the document]
Comment (DeeGeeGit) [reply daemon/services/collectors/notification.go:258]: As per coderabbit suggestions, adopted in 7a3beff: the epoch and mtime-fallback paths now store UTC, so every path renders the same Z-suffixed form. Left the legacy-datetime path as-is (already UTC, and its behaviour is pinned in the PR notes). Storing UTC also made the local-year guard redundant, so that's gone, and the test now pins the literal Z form. Thanks!

File: daemon/services/controllers/notification.go [hunk 2]
```diff
@@ -93,7 +103,16 @@ func ArchiveNotification(id string) error {
...
+	// Since the issue #134 fix an archive copy (without link) is written at
+	// creation, so archiving normally only needs to remove the unread file,
+	// like the stock notify script's 'archive' verb; renaming over the copy
+	// would clobber it. Fall back to a rename for files with no archive copy.
+	if _, err := os.Stat(dst); err == nil {
+		if err := os.Remove(src); err != nil {
+			logger.Error("Failed to archive notification %s: %v", id, err)
+			return fmt.Errorf("failed to archive notification: %w", err)
+		}
+	} else if err := os.Rename(src, dst); err != nil {
```
[the remark this answers is not in the document]
Comment (DeeGeeGit) [reply daemon/services/controllers/notification.go]: Done in 7a3beff — pulled it out as `archiveOrRemove`, with each call site keeping its own error handling. Thanks!

File: daemon/services/controllers/notification.go [hunk 3]
```diff
@@ -186,22 +196,85 @@ func ArchiveAllNotifications() error {
...
+// issue #134 fix an archive copy (without link) is written at creation, so
+// archiving normally only needs to remove the unread file, like the stock
+// notify script's 'archive' verb; renaming over the copy would clobber it.
+// Files with no creation copy (legacy format) fall back to the rename.
+func archiveOrRemove(src, dst string) error {
+	if _, err := os.Stat(dst); err == nil {
+		return os.Remove(src)
+	}
+	return os.Rename(src, dst)
 }
```
[the remark this answers is not in the document]
Comment (DeeGeeGit) [reply daemon/services/controllers/notification.go:286]: Makes sense, but have tweaked it in 1797874 so that on a stat error other than not-exists the helper returns the error rather than removing src. If the check itself failed we can't know whether the copy exists, and removing the unread file in that state could delete the only copy, which felt like the riskier direction. So: definite hit removes, definite miss renames, anything else errors out and leaves both files alone (with a test pinning that no rename or remove is attempted). Thanks!

**Comment by DeeGeeGit** [comment]
Quick fix on the silent temp cleanup, added in 956fff7: a failed removal now logs a warning while the original error is still returned. Thanks!

File: daemon/services/controllers/notification.go [hunk 4]
```diff
@@ -186,22 +196,92 @@ func ArchiveAllNotifications() error {
...
+// so a failed write (e.g. ENOSPC) never leaves a partial or empty .notify
+// file for the Unraid UI to flag as invalid. The temporary name does not end
+// in .notify, keeping it invisible to notification consumers.
+func writeFileAtomic(dir, name string, content []byte) error {
+	tmp, err := os.CreateTemp(dir, ".notify-tmp-*")
+	if err != nil {
+		return err
+	}
+	tmpName := tmp.Name()
+	_, err = tmp.Write(content)
```
[the remark this answers is not in the document]
Comment (DeeGeeGit) [reply daemon/services/controllers/notification.go]: Added in 2410aeb. Worth noting the io.Writer contract already requires an error whenever n < len(p), so os.File can't short-write silently, but os.WriteFile itself carries the same belt-and-braces guard, so happy to match the stdlib here. Thanks!
